Add native Clarity fuzzing for sBTC contracts

moodmosaic opened the pull request:
This adds native, Clarity-based fuzzing for the sBTC contracts, inspired by prior work in [stacks-core#4550](https://github.com/stacks-network/stacks-core/pull/4550) and [stacks-core#4842](https://github.com/stacks-network/stacks-core/pull/4842). 

The approach is similar to Echidna and Foundry but uses *Rendezvous*, a forthcoming tool for Clarity fuzzing. This PR will move out of draft once Rendezvous is publicly available and all sBTC contracts are covered.

BowTiedRadone committed:
Add invariants for `sbtc-token`

(cherry picked from commit a61a4269a6ee58510a847421ea2256f6c0205571)

BowTiedRadone committed:
Add property tests for `sbtc-token`

(cherry picked from commit 7686ffdeca8e13e3b004332b4b9dcccf0baa5339)

moodmosaic committed:
Add placeholders for fuzzing and testing files

- Create `.invariants.clar` files for stateful, invariant fuzzing.
- Create `.tests.clar` files for property-based testing.
- Add placeholder comments to each file indicating their purpose.

BowTiedRadone committed:
Add invariants for the registry contract

(cherry picked from commit d626f47cf3de5d04539dedda6eabb193a7f28c22)

BowTiedRadone committed:
Add four `sbtc-registry` invariants

This commit adds four invariants that check general truths in the registry
contract regarding:
- current-signature-threshold variable
- current-signer-principal variable
- current-aggregate-pubkey variable
- multi-sig-address map

BowTiedRadone committed:
Add property tests for `sbtc-registry`

BowTiedRadone committed:
Unify `invariants` and `property-based` tests

This commit unifies the invariants and the property-based tests for each
contract in a single `.tests` Clarity file.

BowTiedRadone committed:
Add property-based tests for `create-withdrawal-request`

BowTiedRadone committed:
Add dust amount test for `initiate-withdrawal-request`

moodmosaic commented:
Taking the PR out of draft state now that `rendezvous` has been open-sourced (see https://github.com/hirosystems/clarinet/issues/398#issuecomment-2563733568).

djordon commented:
This PR doesn't seem to include any of the necessary testing/fuzzing logic. @moodmosaic Do you plan on adding that here?

moodmosaic commented:
This PR actually does that logic through the `.tests.clar` files. [Rendezvous](https://stacks-network.github.io/rendezvous/chapter_2.html) (`rv`) works differently from traditional fuzzers - it uses the test files we've added to perform both property-based testing and invariant testing.

To clarify:

1. The `.tests.clar` files contain both invariants (e.g., `invariant-signers-always-protocol-caller`) and property tests (e.g., `test-withdrawal-req-id-incremented`). These declarations are what `rv` uses as the specification for fuzzing.

2. `rv` specifically fuzzes at the invariant level to help/make sure contract state remains valid under various conditions. This aims to detect state consistency issues early. Consider the PoX-2 stack-increase bug - while `rv` didn't exist then, tools like this aim to reduce the risk of such critical bugs. This [example](https://github.com/hirosystems/clarity-examples/pull/30) demonstrates this approach.

3. One reason we opened this PR is to test `rv` with real codebases. It's working probabilistically for now - we don't exhaustively explore all possible cases (though we plan to support that via [MiniZinc models](https://github.com/stacks-network/rendezvous/issues/135)).

4. Rendezvous also supports traditional function-level fuzzing to prevent panics and unhandled errors, though without coverage information it's not as efficient as LLVM/clang fuzzing tools. We have an [open issue](https://github.com/stacks-network/rendezvous/issues/85) to help move forward with this.

For context, this technique has already proven valuable - a few issues in PoX-4 were identified and addressed before it went on mainnet. The Rust equivalent would be something like adding [madhouse-rs](https://github.com/stacks-network/madhouse-rs/blob/master/README.md) for stateful fuzzing, as shown in https://github.com/stacks-sbtc/sbtc/pull/1633.

Note that we didn't complete all `.tests.clar` files in this PR as we didn't receive feedback at the time and had to prioritize other work. If by fuzzing you meant something else or expected a different approach, happy to clarify or adjust the PR as needed. 🙏

BowTiedRadone committed:
Merge branch 'stacks-sbtc:main' into rendezvous

BowTiedRadone committed:
Add `@stacks/rendezvous` to `package.json`

BowTiedRadone committed:
Update `sbtc-registry` test contract to compile with `rendezvous`

BowTiedRadone committed:
Update `sbtc-registry` `rendezvous` property tests to current contract logic

BowTiedRadone committed:
Update & cleanup `sbtc-registry` `rendezvous` invariants

BowTiedRadone committed:
Add suggestive comments for `sbtc-registry` `rendezvous` tests

BowTiedRadone committed:
Update `sbtc-token` test contract to compile with `rendezvous`

BowTiedRadone committed:
Update & cleanup `sbtc-token` `rendezvous` invariants

BowTiedRadone committed:
Update `sbtc-token` `rendezvous` property tests to current contract logic

BowTiedRadone committed:
Refine `sbtc-withdrawal` `rendezvous` property tests

BowTiedRadone committed:
Remove unnecessary constant in `sbtc-withdrawal` `rendezvous` tests

BowTiedRadone committed:
Consistency updates in `sbtc-token` `rendezvous` tests

BowTiedRadone committed:
Code cleanup in `sbtc-registry` `rendezvous` tests

setzeus commented:
Hi hi gentlemen! Deeply appreciative of the hard work here - thank you for that penultimate comment with testing instructions. 

Giving it a whirl this weekend & will come back with additional thoughts / requests (or a good ole' LGTM).

setzeus commented:
Hi @moodmosaic @BowTiedRadone, I tried running the tests following the instructions above & get the following:

`setzeus@MacBookPro contracts % npx rv . sbtc-registry invariant
npm ERR! could not determine executable to run

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/setzeus/.npm/_logs/2025-06-28T13_30_26_852Z-debug-0.log
setzeus@MacBookPro contracts % npx rv . sbtc-registry test
npm ERR! could not determine executable to run

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/setzeus/.npm/_logs/2025-06-28T13_30_41_806Z-debug-0.log`

To clarify, I did the following:
1. git clone https://github.com/moodmosaic/sbtc.git
2. cd sbtc/contracts
3. npm i
4. npx rv . sbtc-registry invariant

Am I doing something wrong?

BowTiedRadone commented:
Checked on a clean VM -- no globally installed `rendezvous`. The following steps worked on my end:
1. `git clone https://github.com/moodmosaic/sbtc.git`
2. `cd sbtc/contracts/`
3. `git checkout rendezvous`
4. `npm i`
5. `npx rv . sbtc-registry invariant`

Sample output:
<img width="1464" alt="Screenshot 2025-06-28 at 5 13 27 PM" src="https://github.com/user-attachments/assets/7479194c-42ff-4b2c-ac76-99e23788bd7b" />

setzeus commented:
Hi hi @BowTiedRadone, tried the instructions above & got the following now:

`setzeus@MacBookPro contracts % npx rv . sbtc-registry invariant
/Users/setzeus/Documents/github/sbtc-v2/jul-6-882-review/sbtc/contracts/node_modules/@stacks/rendezvous/node_modules/@hirosystems/clarinet-sdk-wasm/clarinet_sdk.js:1749
const wasmModule = new WebAssembly.Module(bytes);
                   ^

CompileError: WebAssembly.Module(): invalid value type 'externref', enable with --experimental-wasm-reftypes @+109
    at Object.<anonymous> (/Users/setzeus/Documents/github/sbtc-v2/jul-6-882-review/sbtc/contracts/node_modules/@stacks/rendezvous/node_modules/@hirosystems/clarinet-sdk-wasm/clarinet_sdk.js:1749:20)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Module.require (node:internal/modules/cjs/loader:1100:19)
    at require (node:internal/modules/cjs/helpers:119:18)
    at Object.<anonymous> (/Users/setzeus/Documents/github/sbtc-v2/jul-6-882-review/sbtc/contracts/node_modules/@stacks/rendezvous/node_modules/@hirosystems/clarinet-sdk/dist/cjs/node/src/index.js:38:29)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
`

BowTiedRadone commented:
@setzeus that's interesting. I just tried it on my MacBook and it's working fine for me. What's your setup like (Node version, etc.)? Also, here's the Rendezvous [workflow file](https://github.com/stacks-network/rendezvous/blob/master/.github/workflows/ci.yml) we run for all PRs and releases, and all the tests [are passing](https://github.com/stacks-network/rendezvous/actions/runs/15734341250) on macOS in our CI. I'm thinking this might be a versioning issue.

moodmosaic commented:
>I'm thinking this might be a versioning issue.

In order to say, @setzeus can you let us know your environment please? /cc @hugocaillard

djordon-stacks commented:
Closing since we haven't seen much action pushing it through, and we are unlikely to see that anytime soon. Plus, we can always reopen it later.
